File: code/utils/graphs.ts
import { Edge } from "./definitions";
import { stringify } from "querystring";


export interface CreateGraphResult {
    graphs: Map<string, Set<string>>;
    nodes: Set<string>;
}

export default function createDirectedGraph(edges: Edge[], includeNodeSet: boolean = false): CreateGraphResult {
    const graphs = new Map<string, Set<string>>();
    const nodes = new Set<string>();

    for (const edge of edges) {
        if (includeNodeSet) {
            nodes.add(edge.from);
            nodes.add(edge.to);
        }
        if (graphs.has(edge.from)) {
            const set = graphs.get(edge.from);
            if (set !== undefined) {
                set.add(edge.to);
            }
        } else {
            const set = new Set<string>();
            set.add(edge.to);
            graphs.set(edge.from, set);
        }
    }
    return { graphs, nodes };
}

export function createUndirectedGraph(edges: Edge[], includeNodeSet: boolean = false): CreateGraphResult {
    const graphs = new Map<string, Set<string>>();
    const nodes = new Set<string>();

    for (const edge of edges) {
        if (includeNodeSet) {
            nodes.add(edge.from);
            nodes.add(edge.to);
        }

        if (graphs.has(edge.from)) {
            const set = graphs.get(edge.from);
            if (set !== undefined) {
                set.add(edge.to);
            }
        } else {
            const set = new Set<string>();
            set.add(edge.to);
            graphs.set(edge.from, set);
        }

        if (graphs.has(edge.to)) {
            const set = graphs.get(edge.to);
            if (set !== undefined) {
                set.add(edge.from);
            }
        } else {
            const set = new Set<string>();
            set.add(edge.from);
            graphs.set(edge.to, set);
        }
    }
    return { graphs, nodes };
}

File: code/graphs.py
from definitions import Edge

class CreateGraphResult:
    def __init__(self, graphs, nodes):
        self.graphs = graphs
        self.nodes = nodes

def createDirectedGraph(edges, includeNodeSet=False):
    graphs = dict()
    nodes = set()

    for edge in edges:
        if includeNodeSet:
            if edge.src not in nodes:
                nodes.add(edge.src)
            if edge.dest not in nodes:
                nodes.add(edge.dest)

        if edge.src in graphs:
            src_set = graphs[edge.src]
            if src_set is not None:
                src_set.add(edge.dest)
        else:
            new_set = set()
            new_set.add(edge.dest)
            graphs[edge.src] = new_set

    return CreateGraphResult(graphs, nodes)

def createUndirectedGraph(edges, includeNodeSet=False):
    graphs = dict()
    nodes = set()

    for edge in edges:
        if includeNodeSet:
            if edge.src not in nodes:
                nodes.add(edge.src)
            if edge.dest not in nodes:
                nodes.add(edge.dest)

        if edge.src in graphs:
            src_set = graphs[edge.src]
            if src_set is not None:
                src_set.add(edge.dest)
        else:
            new_set = set()
            new_set.add(edge.dest)
            graphs[edge.src] = new_set

        if edge.dest in graphs:
            dest_set = graphs[edge.dest]
            if dest_set is not None:
                dest_set.add(edge.src)
        else:
            new_set = set()
            new_set.add(edge.src)
            graphs[edge.dest] = new_set

    return CreateGraphResult(graphs, nodes)

# result = createUndirectedGraph([
#     Edge('1', '2'),
#     Edge('2', '3'),
#     Edge('1', '3')
# ])
        
# print(result)